tmol/score/dunbrack/potentials/dunbrack_pose_score.impl.hh

```diff
@@ -0,0 +1,384 @@
+
+#include <tmol/utility/tensor/TensorAccessor.h>
+#include <tmol/utility/tensor/TensorPack.h>
+#include <tmol/utility/tensor/TensorStruct.h>
+#include <tmol/utility/tensor/TensorUtil.h>
+#include <tmol/utility/nvtx.hh>
+
+#include <tmol/score/unresolved_atom.hh>
+#include <tmol/score/common/accumulate.hh>
+#include <tmol/score/common/count_pair.hh>
+#include <tmol/score/common/data_loading.hh>
+#include <tmol/score/common/diamond_macros.hh>
+#include <tmol/score/common/geom.hh>
+#include <tmol/score/common/launch_box_macros.hh>
+#include <tmol/score/common/tuple.hh>
+#include <tmol/score/common/uaid_util.hh>
+#include <tmol/score/common/warp_segreduce.hh>
+#include <tmol/score/common/warp_stride_reduce.hh>
+
+#include <tmol/score/dunbrack/potentials/dunbrack_pose_score.hh>
+
+// Operator definitions; safe for CPU compilation
+#include <moderngpu/operators.hxx>
+
+#include <chrono>
+
+#include "potentials.hh"
+
+namespace tmol {
+namespace score {
+namespace dunbrack {
+namespace potentials {
+
+template <typename Real, int N>
+using Vec = Eigen::Matrix<Real, N, 1>;
+
+template <
+    template <tmol::Device>
+    class DeviceDispatch,
+    tmol::Device D,
+    typename Real,
+    typename Int>
+auto DunbrackPoseScoreDispatch<DeviceDispatch, D, Real, Int>::f(
+    TView<Vec<Real, 3>, 2, D> coords,
+    TView<Int, 2, D> pose_stack_block_coord_offset,
+    TView<Int, 2, D> pose_stack_block_type,
+    TView<Vec<Int, 2>, 3, D> pose_stack_inter_block_connections,
+    TView<Int, 3, D> block_type_atom_downstream_of_conn,
+
+    TView<Real, 3, D> rotameric_neglnprob_tables,
+    TView<Vec<int64_t, 2>, 1, D> rotprob_table_sizes,
+    TView<Vec<int64_t, 2>, 1, D> rotprob_table_strides,
+    TView<Real, 3, D> rotameric_mean_tables,
+    TView<Real, 3, D> rotameric_sdev_tables,
+    TView<Vec<int64_t, 2>, 1, D> rotmean_table_sizes,
+    TView<Vec<int64_t, 2>, 1, D> rotmean_table_strides,
+
+    TView<Vec<Real, 2>, 1, D> rotameric_bb_start,        // ntable-set entries
+    TView<Vec<Real, 2>, 1, D> rotameric_bb_step,         // ntable-set entries
+    TView<Vec<Real, 2>, 1, D> rotameric_bb_periodicity,  // ntable-set entries
+
+    TView<Int, 1, D> rotameric_rotind2tableind,
+    TView<Int, 1, D> semirotameric_rotind2tableind,
+
+    TView<Real, 4, D> semirotameric_tables,              // n-semirot-tabset
+    TView<Vec<int64_t, 3>, 1, D> semirot_table_sizes,    // n-semirot-tabset
+    TView<Vec<int64_t, 3>, 1, D> semirot_table_strides,  // n-semirot-tabset
+    TView<Vec<Real, 3>, 1, D> semirot_start,             // n-semirot-tabset
+    TView<Vec<Real, 3>, 1, D> semirot_step,              // n-semirot-tabset
+    TView<Vec<Real, 3>, 1, D> semirot_periodicity,       // n-semirot-tabset
+
+    TView<Int, 1, D> block_n_dihedrals,
+    TView<UnresolvedAtomID<Int>, 2, D> block_phi_uaids,
+    TView<UnresolvedAtomID<Int>, 2, D> block_psi_uaids,
+    TView<UnresolvedAtomID<Int>, 3, D>
+        block_chi_uaids,  // TODO: unused, can probably be removed
```

Comment:
Andrew can comment on this.  In theory a chi could cross a residue boundary ... but the code would need to be rewritten for that purpose...

Reply:
Ah, I forgot I had a couple TODOs in here.

I made the note because all of the chis are encoded in the block_dih_uaids (arg right below this). I originally separated them out, but it was easier to work with existing code by pooling them together. I'll go ahead and remove this unless there's a reason to keep it.

Side note: would it be beneficial to stick an extra check in our PR testing that looks for TODOs?